### PR TITLE
Implement configuration lookup from env variables

### LINKED ISSUE #48: Implement configuration lookup from environment variable
From the current approach, when we need to read the configurations from environment variables, we need to specifically mention in the deployment.yaml file as
> gateWayEndpoint: ${env:gateWayEndpoint}

But in a containerized environment, changing configuration file is harder, because we cannot maintain a single Container Image for different deployments, and manipulate the Container runtimes via environment variables.

From the suggested approach, read priority will be,
1. Read configurations from environment variables. environment variable should define in a certain format like
> PREFIX_wso2.NAMESPACE_CONFIGKEY="value", PREFIX_wso2.NAMESPACE_CONFIGPARENT_CONFIG2="value2"
2. Read the deployment.yaml file.
3. Get default values from the bean class.

### LINKED ISSUE #51: Update documentation for configuration lookup from environment variable
Necessary documentation changes for issue #48

### PR DESCRIPTION
**Pull request by thusithathilina**
Resolves #48 and Resolves #51

### CODE REVIEW

File: components/org.wso2.carbon.config/src/main/java/org/wso2/carbon/config/provider/ImmutablePair.java [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+ *   Tuple t = tupleStreamPair.getFirst();
+ *   TupleInputStreamIdentifier identifier = tupleStreamPair.getSecond();
+ *   ...
+ * </pre>
+ *
+ * @param <A> type A
+ * @param <B> type B
+ * @since 2.0.6
+ */
+final class ImmutablePair<A, B> {
```
Comment (Kishanthan) [inline components/org.wso2.carbon.config/src/main/java/org/wso2/carbon/config/provider/ImmutablePair.java]: Whats the purpose of the class and who uses this?

File: components/org.wso2.carbon.config/src/main/java/org/wso2/carbon/config/provider/ConfigProviderImpl.java [hunk 2]
```diff
@@ -136,6 +153,438 @@ public Object getConfigurationObject(String namespace) throws ConfigurationExcep
...
     }
 
+    /**
+     * Returns a map of variables which are prefixed with the given namespace. A hash map will always be returned
+     * even if no variables are found which are prefixed with the given namespace.
+     * <p>
+     * A variable should at least consists of {@value CONFIG_MIN_ELEMENT_COUNT} elements, the configuration namespace
+     * and the configuration element.
+     *
+     * @param namespace configuration namespace
```
Comment (ChanakaCooray) [inline components/org.wso2.carbon.config/src/main/java/org/wso2/carbon/config/provider/ConfigProviderImpl.java]: missing Javadoc for variables.

File: components/org.wso2.carbon.config/src/main/java/org/wso2/carbon/config/provider/ImmutablePair.java (hunk 1 above)
Comment (vidurananayakkara) [reply components/org.wso2.carbon.config/src/main/java/org/wso2/carbon/config/provider/ImmutablePair.java]: This is used in [1], where we return the unique attribute field name and its value.

Ex:- 
```
TESTCONFIGURATION_TRANSPORTS_TRANSPORT_0_NAME="abc"
TESTCONFIGURATION_TRANSPORTS_TRANSPORT_0_PORT="8000"
```

In here, the unique identifier's env variable is `TESTCONFIGURATION_TRANSPORTS_TRANSPORT_0_NAME`. If you pass `TESTCONFIGURATION_TRANSPORTS_TRANSPORT_0_PORT` in the method [1], you will get ImmutablePair.of("NAME", "abc")

We do need the name and the value when we need to identify whether a configuration needs to be overridden or added (ex: add a new transport via environment variables)

[1] https://github.com/wso2/carbon-config/pull/54/files#diff-a47faff6be5f4d2f88ebc8088d7a1c29R380

File: components/org.wso2.carbon.config/src/test/java/org/wso2/carbon/config/utils/EnvironmentUtils.java [hunk 3]
```diff
@@ -60,31 +60,52 @@ public static void setEnv(Map<String, String> newVariables) {
         Map<String, String> newenv = new HashMap<>();
```
Comment (Kishanthan) [inline components/org.wso2.carbon.config/src/test/java/org/wso2/carbon/config/utils/EnvironmentUtils.java]: Lets rename this to setEnvironementVariables as this is a public method

Comment (thusithathilina) [reply components/org.wso2.carbon.config/src/test/java/org/wso2/carbon/config/utils/EnvironmentUtils.java]: changed as requested

File: components/org.wso2.carbon.config/src/test/java/org/wso2/carbon/config/utils/EnvironmentUtils.java [hunk 4]
```diff
@@ -45,46 +45,67 @@
...
             Field theEnvironmentField = processEnvironmentClass.getDeclaredField(THE_ENVIRONMENT_FILED);
             theEnvironmentField.setAccessible(true);
             Map<String, String> env = (Map<String, String>) theEnvironmentField.get(null);
-            env.putAll(newenv);
+            env.putAll(environmentVariables);
 
             Field theCaseInsensitiveEnvironmentField = processEnvironmentClass
                     .getDeclaredField(THE_CASE_INSENSITIVE_ENVIRONMENT);
             theCaseInsensitiveEnvironmentField.setAccessible(true);
             Map<String, String> cienv = (Map<String, String>) theCaseInsensitiveEnvironmentField.get(null);
```
Comment (Kishanthan) [inline components/org.wso2.carbon.config/src/test/java/org/wso2/carbon/config/utils/EnvironmentUtils.java]: Use proper variable names here please. What is "cienv" ?

**Pull request merged**